cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringProgramEventPublisher.java

```diff
@@ -0,0 +1,232 @@
+      if (lastProgramUpdateMessageId == null) {
+        // Initialize subscriber based on last message fetched by RuntimeProgramStatusSubscriberService.
+        // This is to avoid fetching existing program history from before TetheringAgent was added
+        String messageId = appMetadataStore.retrieveSubscriberState(programUpdateTopic,
+                                                                    RuntimeProgramStatusSubscriberService.SUBSCRIBER);
+        appMetadataStore.persistSubscriberState(programUpdateTopic, TetheringAgentService.SUBSCRIBER, messageId);
+        lastProgramUpdateMessageId = messageId;
+      }
+    });
+  }
+
+  @Override
+  protected long runTask() throws IOException {
+    List<PeerInfo> peers = store.getPeers().stream()
+      // Ignore peers in REJECTED state.
+      .filter(p -> p.getTetheringStatus() != TetheringStatus.REJECTED)
+      .collect(Collectors.toList());
+
+    PeerProgramUpdates peerProgramUpdates = getPeerProgramUpdates();
+    persistNotifications(peerProgramUpdates, peers);
+    TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+      appMetadataStore.persistSubscriberState(programUpdateTopic,
+                                              TetheringAgentService.SUBSCRIBER,
+                                              lastProgramUpdateMessageId);
+
+    });
+
+    return pollInterval;
+  }
+
+  /**
+   * Returns program status updates for tethered peers.
+   */
+  private PeerProgramUpdates getPeerProgramUpdates() {
+    Map<String, List<Notification>> peerToNotifications = new HashMap<>();
+    String lastMessageId = lastProgramUpdateMessageId;
+    try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                    programUpdateTopic,
+                                                                    programUpdateFetchSize,
+                                                                    lastProgramUpdateMessageId)) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
+        if (notification.getNotificationType() == Notification.Type.PROGRAM_STATUS) {
+          Map<String, String> properties = notification.getProperties();
+          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
+          try {
+            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
+            if (runRecord.getPeerName() != null) {
+              String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
+              LOG.debug("Notifying peer {} about program run {} in state {}",
```

Comment:
this isn't really an accurate log message, as we could fail before we get to the part where we contact the peer. The message should be something about receiving a message.

It's also more of a trace message, as we don't want a debug message for every state transition.

Reply:
I changed the log message to "Received message for peer {} about program run {} in state {}".

Prefer to keep it at debug log level as it's useful for debugging issues where tethered program state doesn't get updated correctly (it's helpful to understand if a missing program state update was not generated not on client side, or dropped/handled incorrectly on the server side). These issues are usually not readily reproducible, so it helps to have  logs available when they do happen.
WDYT?

Reply:
hm ok we can keep it